test/test_multidevice_collectives.cpp

```diff
@@ -0,0 +1,242 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#ifdef USE_DISTRIBUTED
+#include <gtest/gtest.h>
+
+#include <multidevice/collective.h>
+#include <multidevice/communicator.h>
+#include <test/utils.h>
+
+#include <iostream>
+
+namespace nvfuser {
+
+static constexpr DeviceIdxType root = 0;
+static constexpr int tensor_size = 1024;
+static constexpr int number_of_repetitions = 8;
+
+TEST_F(NVFuserTest, FusionMultiGPU_Collective_Gather_CUDA) {
+  if (!comm.is_available() || comm.size() < 2) {
+    GTEST_SKIP() << "This test needs at least 2 ranks";
+  }
+  c10::TensorOptions options =
+      at::TensorOptions().dtype(at::kFloat).device(comm.device());
+
+  CommParams params;
+  params.root = root;
+  params.team = std::vector<DeviceIdxType>(comm.size());
+  std::iota(params.team.begin(), params.team.end(), 0);
+  params.src_bufs = {at::ones(tensor_size, options) * comm.deviceId()};
+  if (comm.deviceId() == root) {
+    for (int i = 0; i < comm.size(); i++) {
+      params.dst_bufs.push_back(-at::ones(tensor_size, options));
+    }
+  }
+  for (int i = 0; i < number_of_repetitions; i++) {
```

Comment:
What do you intend to test with this repetition?

Shouldn't we clear the destination buffers at each iteration? Otherwise, we might miss some synchronization bugs.

Reply:
Changes in the last commit:
- the collective is instantiated outside of the iteration loop (this was a mistake before)
- at each iteration, the source buffers are initialized to different values each time
- at each iteration, destination buffers are initialized to 0